### PR TITLE
Stage4

### PR DESCRIPTION
**Pull request by KateMoreva**

### CODE REVIEW

**Comment by KateMoreva** [comment]
Результаты нагрузочного тестирования будут в ближайшее время.

**Comment by KateMoreva** [comment]
### PUT запросы 

-t1 -c64 -d60s -R2000 -s ./wrk/put.lua --latency http://localhost:8081

**CPU**


Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.467ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.42ms  679.72us   9.50ms   68.92%
    Req/Sec     2.11k   249.84     3.44k    69.71%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.35ms
 75.000%    1.86ms
 90.000%    2.30ms
 99.000%    2.86ms
 99.900%    6.32ms
 99.990%    8.33ms
 99.999%    9.19ms
100.000%    9.51ms
  119713 requests in 1.00m, 9.48MB read
Requests/sec:   1995.17
Transfer/sec:    161.72KB

**Alloc**

Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.520ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.47ms  705.08us  10.63ms   68.84%
    Req/Sec     2.11k   260.26     3.56k    67.12%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.40ms
 75.000%    1.87ms
 90.000%    2.41ms
 99.000%    3.00ms
 99.900%    6.43ms
 99.990%    8.04ms
 99.999%    8.70ms
100.000%   10.64ms
  119714 requests in 1.00m, 9.48MB read
Requests/sec:   1995.19
Transfer/sec:    161.72KB

**Lock**

Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.421ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.40ms  661.70us   9.49ms   68.12%
    Req/Sec     2.10k   252.82     3.44k    69.54%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.32ms
 75.000%    1.84ms
 90.000%    2.24ms
 99.000%    2.78ms
 99.900%    6.08ms
 99.990%    7.77ms
 99.999%    8.57ms
100.000%    9.49ms
  119714 requests in 1.00m, 9.48MB read
Requests/sec:   1995.19
Transfer/sec:    161.72KB

### GET запросы 

-t1 -c64 -d60s -R2000 -s ./wrk/get.lua --latency http://localhost:8081

**CPU**


Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.398ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.39ms  661.69us   8.61ms   68.09%
    Req/Sec     2.10k   210.07     3.20k    71.62%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.31ms
 75.000%    1.80ms
 90.000%    2.27ms
 99.000%    2.85ms
 99.900%    5.70ms
 99.990%    7.13ms
 99.999%    8.06ms
100.000%    8.61ms
  119715 requests in 1.00m, 10.16MB read
Requests/sec:   1995.20
Transfer/sec:    173.39KB

**Alloc**

Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.464ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.44ms  706.00us  12.92ms   68.59%
    Req/Sec     2.11k   212.65     3.44k    72.73%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.37ms
 75.000%    1.85ms
 90.000%    2.42ms
 99.000%    3.02ms
 99.900%    5.72ms
 99.990%    7.89ms
 99.999%   12.68ms
100.000%   12.93ms
  119714 requests in 1.00m, 10.16MB read
Requests/sec:   1995.22
Transfer/sec:    173.39KB

**Lock**

Running 1m test @ http://localhost:8081
  1 threads and 64 connections
  Thread calibration: mean lat.: 1.330ms, rate sampling interval: 10ms
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.33ms  661.62us  10.17ms   67.56%
    Req/Sec     2.10k   269.42     3.40k    65.57%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.27ms
 75.000%    1.72ms
 90.000%    2.26ms
 99.000%    2.78ms
 99.900%    5.44ms
 99.990%    8.02ms
 99.999%   10.11ms
100.000%   10.18ms
  119708 requests in 1.00m, 10.16MB read
Requests/sec:   1995.14
Transfer/sec:    173.38KB

Для начала было проведено тестирование с достаточно низкой нагрузкой 2000R.
Тестирование было проведено на ноде 8081.

**Comment by KateMoreva** [comment]
### PUT Запросы 

-t64 -c64 -d60s -R40000 -s ./wrk/put.lua --latency http://localhost:8080

**CPU**


Running 1m test @ http://localhost:8080
  64 threads and 64 connections
Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   652.66us  332.60us   9.02ms   65.69%
    Req/Sec   674.73     45.53     1.11k    77.32%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  645.00us
 75.000%    0.90ms
 90.000%    1.05ms
 99.000%    1.22ms
 99.900%    3.19ms
 99.990%    5.49ms
 99.999%    7.26ms
100.000%    9.03ms
  2399319 requests in 1.00m, 189.92MB read
Requests/sec:  40033.62
Transfer/sec:      3.17MB

**Alloc**


Running 1m test @ http://localhost:8080
  64 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   665.96us  368.07us  17.34ms   72.26%
    Req/Sec   672.65     47.64     1.78k    75.35%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  653.00us
 75.000%    0.91ms
 90.000%    1.06ms
 99.000%    1.50ms
 99.900%    3.64ms
 99.990%    8.57ms
 99.999%   13.05ms
100.000%   17.36ms
  2399290 requests in 1.00m, 189.92MB read
Requests/sec:  40031.51
Transfer/sec:      3.17MB

**Lock**


Running 1m test @ http://localhost:8080
  64 threads and 64 connections
    Latency   652.15us  344.20us  12.52ms   68.10%
    Req/Sec   675.36     45.63     1.33k    77.86%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  643.00us
 75.000%    0.90ms
 90.000%    1.05ms
 99.000%    1.21ms
 99.900%    3.36ms
 99.990%    7.35ms
 99.999%   10.50ms
100.000%   12.53ms
  2399325 requests in 1.00m, 189.92MB read
Requests/sec:  40036.03
Transfer/sec:      3.17MB

**Comment by KateMoreva** [comment]
### GET запросы 

**CPU**

  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   657.85us  338.07us  11.99ms   66.76%
    Req/Sec   672.96     46.36     1.22k    76.05%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  649.00us
 75.000%    0.90ms
 90.000%    1.05ms
 99.000%    1.22ms
 99.900%    3.31ms
 99.990%    6.45ms
 99.999%    8.35ms
100.000%   12.00ms
  2399348 requests in 1.00m, 202.36MB read
Requests/sec:  40025.90
Transfer/sec:      3.38MB

**Alloc**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   660.67us  342.45us  12.63ms   67.58%
    Req/Sec   671.54     46.89     1.33k    75.07%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  652.00us
 75.000%    0.91ms
 90.000%    1.06ms
 99.000%    1.24ms
 99.900%    3.43ms
 99.990%    6.52ms
 99.999%    9.60ms
100.000%   12.64ms
  2399298 requests in 1.00m, 202.36MB read
Requests/sec:  40038.45
Transfer/sec:      3.38MB

**Lock**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   655.20us  329.65us  10.66ms   65.21%
    Req/Sec   673.46     45.94     1.22k    76.44%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  649.00us
 75.000%    0.90ms
 90.000%    1.05ms
 99.000%    1.21ms
 99.900%    2.91ms
 99.990%    6.11ms
 99.999%    7.98ms
100.000%   10.67ms
  2398436 requests in 1.00m, 202.28MB read
Requests/sec:  40020.46
Transfer/sec:      3.38MB

**Comment by vaddya** [comment]
Профилируйте дальше без `-t`, пожалуйста, а то тяжело флеймграфы читать

**Comment by KateMoreva** [comment]
> Профилируйте дальше без `-t`, пожалуйста, а то тяжело флеймграфы читать

Хорошо, дальше буду использовать количество потоков по умолчанию.

**Comment by vaddya** [comment]
> > Профилируйте дальше без `-t`, пожалуйста, а то тяжело флеймграфы читать
> 
> Хорошо, дальше буду использовать количество потоков по умолчанию.

Это, если мне память не изменяет, про то, чтобы не разделять на флеймграфах потоки, а не про их количество

**Comment by KateMoreva** [comment]
> > > Профилируйте дальше без `-t`, пожалуйста, а то тяжело флеймграфы читать
> > 
> > 
> > Хорошо, дальше буду использовать количество потоков по умолчанию.
> 
> Это, если мне память не изменяет, про то, чтобы не разделять на флеймграфах потоки, а не про их количество

Ой, да это же про профилирование разных потоков по отдельности.

**Review by vaddya — CHANGES_REQUESTED** [review]
Небольшие замечания и вопросы по реализации.

По профилированию:
- Давайте дальше профилировать без разделения на потоки (`-t`) в `async-profiler`, а то сложно анализировать суммарный вклад разных участков кода.
- Можете сравнить результаты профлирования с прошлым этапом? Стало ли быстрее/медленнее? Выросло/упало число запросов, с которым справляется сервер?

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+        this.executorService = new ThreadPoolExecutor(numberOfWorkers,
+                queueSize,
+                0L,
+                TimeUnit.MILLISECONDS,
+                new ArrayBlockingQueue<>(queueSize),
+                new ThreadFactoryBuilder()
+                        .setNameFormat("Worker_%d")
+                        .setUncaughtExceptionHandler((t, e) -> log.error("Error in {} when processing request", t, e))
+                        .build(),
+                new ThreadPoolExecutor.DiscardOldestPolicy());
```
[the remark this answers is not in the document]
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Подумала, что лучше удалять задачу во главе рабочей очереди и повторять попытку выполнения, но, наверное, лучше использовать AbortPolicy, который выдает исключение RejectedExecutionException при отклонении, и при необходимости его обрабатывать.

**Comment by KateMoreva** [comment]
Не проходят лишь тесты 5ого этапа.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        } catch (IOException e) {
+            log.error(SERVER_ERROR, session, e);
+        }
+    }
+
+    private void proxyRequest(final HttpSession session, final Response response) {
+        try {
+            session.sendResponse(response);
+        } catch (IOException e) {
+            log.error(SERVER_ERROR, session, e);
```
[the remark this answers is not in the document]
Comment (KateMoreva) [reply src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java]: Нет, у нее нет адекватного строкового представления, исправила.

**Comment by KateMoreva** [comment]
> Небольшие замечания и вопросы по реализации.
> 
> По профилированию:
> 
> * Давайте дальше профилировать без разделения на потоки (`-t`) в `async-profiler`, а то сложно анализировать суммарный вклад разных участков кода.
> * Можете сравнить результаты профлирования с прошлым этапом? Стало ли быстрее/медленнее? Выросло/упало число запросов, с которым справляется сервер?

Спасибо за замечания по реализации, исправила все.

**Comment by KateMoreva** [comment]
Для сравнения производительности работы кластера была проведена серия тестов с параметрами -t64 -c64 -d60s -R80000.
Кластер больше не держит нагрузку в 80000 заявок в секунду. 

### PUT запросы

**CPU**



Running 1m test @ http://localhost:8080
  64 threads and 64 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     5.84ms   12.64ms  85.12ms   89.44%
    Req/Sec     1.25k    23.94     1.36k    83.69%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.25ms
 75.000%    1.99ms
 90.000%   20.16ms
 99.000%   62.94ms
 99.900%   74.50ms
 99.990%   81.92ms
 99.999%   84.35ms
100.000%   85.18ms
  4797840 requests in 1.00m, 379.78MB read
Requests/sec:  80037.61
Transfer/sec:      6.34MB

**Alloc**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    56.75ms  104.13ms 429.31ms   82.13%
    Req/Sec     1.24k    23.25     1.33k    89.31%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.41ms
 75.000%   45.69ms
 90.000%  260.73ms
 99.000%  356.86ms
 99.900%  412.93ms
 99.990%  427.52ms
 99.999%  429.05ms
100.000%  429.57ms
  4767722 requests in 1.00m, 377.40MB read
Requests/sec:  79531.68
Transfer/sec:      6.30MB

**Lock**



  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   287.46ms  422.82ms   1.46s    79.24%
    Req/Sec     1.22k    44.61     1.34k    83.36%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.56ms
 75.000%  542.21ms
 90.000%    1.04s 
 99.000%    1.36s 
 99.900%    1.44s 
 99.990%    1.46s 
 99.999%    1.46s 
100.000%    1.46s 
  4688625 requests in 1.00m, 371.13MB read
Requests/sec:  78187.42
Transfer/sec:      6.19MB

### GET запросы

**CPU**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   232.86ms  360.93ms   1.21s    80.29%
    Req/Sec     1.22k    43.85     1.30k    87.98%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%    1.62ms
 75.000%  451.84ms
 90.000%  891.90ms
 99.000%    1.14s 
 99.900%    1.18s 
 99.990%    1.20s 
 99.999%    1.21s 
100.000%    1.21s
  4705568 requests in 1.00m, 398.11MB read
Requests/sec:  78557.79
Transfer/sec:      6.65MB

**Alloc**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency   296.65ms  418.38ms   1.45s    79.39%
    Req/Sec     1.22k    50.27     1.43k    77.24%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%   27.07ms
 75.000%  586.24ms
 90.000%    1.02s 
 99.000%    1.34s 
 99.900%    1.41s 
 99.990%    1.43s 
 99.999%    1.45s 
100.000%    1.45s 
  4689203 requests in 1.00m, 396.72MB read
Requests/sec:  78260.56
Transfer/sec:      6.62MB

**Lock**


  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.03s     1.16s    3.47s    76.81%
    Req/Sec     1.17k    81.35     1.45k    64.65%
  Latency Distribution (HdrHistogram - Recorded Latency)
 50.000%  440.83ms
 75.000%    2.11s 
 90.000%    2.87s 
 99.000%    3.29s 
 99.900%    3.42s 
 99.990%    3.45s 
 99.999%    3.46s 
100.000%    3.47s 
  4530971 requests in 1.00m, 383.29MB read
Requests/sec:  75585.72
Transfer/sec:      6.39MB

**Comment by vaddya** [comment]
Секунды для нашего ДАО довольно большие числа, надо уменьшать нагрузку. У меня еще локально фейлится `ShardingTest > distribute`, у вас ОК?

**Comment by KateMoreva** [comment]
> Секунды для нашего ДАО довольно большие числа, надо уменьшать нагрузку. У меня еще локально фейлится `ShardingTest > distribute`, у вас ОК?

Да, сервер не держит такую нагрузку, а вот 40000 для него вполне нормально. 
Странно, у меня локально тест проходит.

**Comment by vaddya** [comment]
> Да, сервер не держит такую нагрузку, а вот 40000 для него вполне нормально.
> Странно, у меня локально тест проходит.

Хм, у меня один раз на 5-6 запусков этот тест фейлится с:
```
execution timed out after 60000 ms
org.opentest4j.AssertionFailedError: execution timed out after 60000 ms
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:144)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:115)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:97)
	at org.junit.jupiter.api.AssertTimeout.assertTimeoutPreemptively(AssertTimeout.java:93)
	at org.junit.jupiter.api.Assertions.assertTimeoutPreemptively(Assertions.java:3236)
	at ru.mail.polis.service.ShardingTest.distribute(ShardingTest.java:227)
```

**Comment by vaddya** [comment]
Ладно, такое ощущение, что это моя тачка барахлит, да и багов не вижу, забиваем

File: src/main/java/ru/mail/polis/service/ServiceFactory.java [hunk 3]
```diff
@@ -54,7 +58,13 @@ public static Service create(
         if (port <= 0 || 65536 <= port) {
             throw new IllegalArgumentException("Port out of range");
         }
-
-        throw new UnsupportedOperationException("Implement me!");
+        final Topology<String> moduleTopology = new ModuleTopology(topology,
+                String.format("http://localhost:%d",
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/ServiceFactory.java:62]: Для такого подойдет обычный `+`.

File: src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public boolean isMe(@NotNull final String node) {
+        return node.equals(me);
+    }
+
+    @NotNull
+    @Override
+    public List<String> all() {
+        return nodes;
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/ModuleTopology.java:52]: Опасная практика -- отдавать наружу изменяемую коллекцию, от которой зависит наша работа. Предлагаю побыть параноиком и делать тут копию.

File: src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            });
+        } catch (RejectedExecutionException e) {
+            handleError(session, Response.SERVICE_UNAVAILABLE);
+        }
+    }
+
+    private void getMethod(final ByteBuffer key, final HttpSession session) {
+        try {
+            getEntity(key, session);
+        } catch (IOException e) {
```
Comment (vaddya) [inline src/main/java/ru/mail/polis/service/kate/moreva/MySimpleHttpServer.java:168]: Давайте словим и залогируем это исключение в методе `entity`? Избавимся от повторений, и методы `getMethod` и `getEntity` можно совместить в один (другие по аналогии).

**Review by vaddya — CHANGES_REQUESTED** [review]
Еще чуть-чуть по реализации.

**Comment by KateMoreva** [comment]
> Еще чуть-чуть по реализации.

Спасибо, исправила, определение вида запроса пришлось все же вынести в отдельный метод, тк код стал нечитаемым, да и климат ругался.

**Review by vaddya — APPROVED** [review]
Принято.

Собирайте, пожалуйста, на следующих этапах результаты профилирования одного этапа в один Markdown файл. Вывод `wrk`, флейм граф от `async-profiler` (прямо в Markdown `get-cpu`) и краткие выводы (то, что у вас сейчас комментами в PR написано). И так для каждой комбинации `get`/`put`, `cpu`/`alloc`/`lock`. Будет гораздо удобнее.

**Comment by KateMoreva** [comment]
> Принято.
> 
> Собирайте, пожалуйста, на следующих этапах результаты профилирования одного этапа в один Markdown файл. Вывод `wrk`, флейм граф от `async-profiler` (прямо в Markdown `get-cpu`) и краткие выводы (то, что у вас сейчас комментами в PR написано). И так для каждой комбинации `get`/`put`, `cpu`/`alloc`/`lock`. Будет гораздо удобнее.

Спасибо, хорошо, буду собирать в один файл.

**Pull request closed without merging**